File: code/packages/go/network-stack/udp.go
package networkstack

// UDP — Layer 4 (Transport), Unreliable Datagrams
//
// UDP is the simpler sibling of TCP. Where TCP is registered mail with
// tracking, UDP is a postcard — fast, simple, and unreliable.
//
// No connection setup, no acknowledgments, no retransmissions. You write
// the message and drop it in the mailbox. It might get lost.
//
// Used for DNS (single question-answer), video streaming (a lost frame is
// better than a delayed frame), and online gaming (newest position update
// is all that matters).
//
// # UDP Header (8 bytes)
//
//	Bytes 0-1: source port
//	Bytes 2-3: destination port
//	Bytes 4-5: length (header + payload)
//	Bytes 6-7: checksum

import (
	"encoding/binary"
	"errors"
)

// UDPHeader represents an 8-byte UDP datagram header.
type UDPHeader struct {
	SrcPort  uint16 // Source port (can be 0 if no reply expected)
	DstPort  uint16 // Destination port
	Length   uint16 // Total datagram size (header + payload), minimum 8
	Checksum uint16 // Error detection (0 = "not computed" in IPv4)
}

// NewUDPHeader creates a header with default values (length=8, ports=0).
func NewUDPHeader() *UDPHeader {
	return &UDPHeader{Length: 8}
}

// Serialize converts the header to 8 raw bytes in network byte order.
func (h *UDPHeader) Serialize() []byte {
	buf := make([]byte, 8)
	binary.BigEndian.PutUint16(buf[0:2], h.SrcPort)
	binary.BigEndian.PutUint16(buf[2:4], h.DstPort)
	binary.BigEndian.PutUint16(buf[4:6], h.Length)
	binary.BigEndian.PutUint16(buf[6:8], h.Checksum)
	return buf
}

// DeserializeUDPHeader parses 8 bytes into a UDPHeader.
func DeserializeUDPHeader(data []byte) (*UDPHeader, error) {
	if len(data) < 8 {
		return nil, errors.New("UDP header too short: minimum 8 bytes")
	}
	return &UDPHeader{
		SrcPort:  binary.BigEndian.Uint16(data[0:2]),
		DstPort:  binary.BigEndian.Uint16(data[2:4]),
		Length:   binary.BigEndian.Uint16(data[4:6]),
		Checksum: binary.BigEndian.Uint16(data[6:8]),
	}, nil
}

// Datagram represents a received UDP datagram with its source address.
type Datagram struct {
	Data    []byte
	SrcIP   uint32
	SrcPort uint16
}

// UDPSocket sends and receives individual datagrams. Unlike TCP, there is
// no connection — each datagram is independent. Think of it like a mailbox:
// you can send to anyone and receive from anyone.
type UDPSocket struct {
	LocalPort uint16
	recvQueue []Datagram
}

// NewUDPSocket creates a UDP socket bound to the given port.
func NewUDPSocket(localPort uint16) *UDPSocket {
	return &UDPSocket{LocalPort: localPort}
}

// SendTo creates a UDP datagram for the given destination.
// Returns the header and payload. The caller (IP layer) handles transmission.
func (s *UDPSocket) SendTo(data []byte, destIP uint32, destPort uint16) (*UDPHeader, []byte) {
	header := &UDPHeader{
		SrcPort: s.LocalPort,
		DstPort: destPort,
		Length:  uint16(8 + len(data)),
	}
	return header, data
}

// ReceiveFrom returns the next datagram from the queue, or nil if empty.
// Unlike TCP's Receive, this returns the sender's address because UDP is
// connectionless.
func (s *UDPSocket) ReceiveFrom() *Datagram {
	if len(s.recvQueue) == 0 {
		return nil
	}
	d := s.recvQueue[0]
	s.recvQueue = s.recvQueue[1:]
	return &d
}

// Deliver places a received datagram into this socket's receive queue.
// This simulates the kernel's job of demultiplexing incoming datagrams.
func (s *UDPSocket) Deliver(data []byte, srcIP uint32, srcPort uint16) {
	s.recvQueue = append(s.recvQueue, Datagram{
		Data:    data,
		SrcIP:   srcIP,
		SrcPort: srcPort,
	})
}

// HasData returns true if there are datagrams waiting to be read.
func (s *UDPSocket) HasData() bool {
	return len(s.recvQueue) > 0
}

File: code/packages/rust/network-stack/src/udp.rs
// ============================================================================
// Layer 4: UDP — User Datagram Protocol
// ============================================================================
//
// UDP is the "anti-TCP" — no reliability, no ordering, no flow control,
// no connection state. You send a datagram; it either arrives or it doesn't.
//
// Why use something so unreliable? Because simplicity and speed matter:
//   - DNS lookups: one question, one answer
//   - Video streaming: a dropped frame is better than pausing
//   - Online games: latest position matters, old ones don't
//
// The entire UDP header is only 8 bytes (vs TCP's 20). Less overhead per
// packet means higher throughput for high-frequency small messages.
//
// UDP Header:
//   +-----------+-----------+--------+----------+
//   | src_port  | dst_port  | length | checksum |
//   | (2 bytes) | (2 bytes) | (2 B)  | (2 B)    |
//   +-----------+-----------+--------+----------+
//
// ============================================================================

use crate::ethernet::Ipv4Address;

// ============================================================================
// UdpHeader — 8-byte datagram header
// ============================================================================
#[derive(Debug, Clone, PartialEq)]
pub struct UdpHeader {
    pub src_port: u16,
    pub dst_port: u16,
    pub length: u16,
    pub checksum: u16,
}

impl UdpHeader {
    pub fn new(src_port: u16, dst_port: u16, length: u16) -> Self {
        Self { src_port, dst_port, length, checksum: 0 }
    }

    /// Serialize into an 8-byte vector.
    pub fn serialize(&self) -> Vec<u8> {
        let mut bytes = vec![0u8; 8];
        bytes[0] = (self.src_port >> 8) as u8;
        bytes[1] = (self.src_port & 0xFF) as u8;
        bytes[2] = (self.dst_port >> 8) as u8;
        bytes[3] = (self.dst_port & 0xFF) as u8;
        bytes[4] = (self.length >> 8) as u8;
        bytes[5] = (self.length & 0xFF) as u8;
        bytes[6] = (self.checksum >> 8) as u8;
        bytes[7] = (self.checksum & 0xFF) as u8;
        bytes
    }

    /// Deserialize from a byte slice.
    pub fn deserialize(bytes: &[u8]) -> Option<Self> {
        if bytes.len() < 8 {
            return None;
        }

        Some(Self {
            src_port: ((bytes[0] as u16) << 8) | (bytes[1] as u16),
            dst_port: ((bytes[2] as u16) << 8) | (bytes[3] as u16),
            length: ((bytes[4] as u16) << 8) | (bytes[5] as u16),
            checksum: ((bytes[6] as u16) << 8) | (bytes[7] as u16),
        })
    }
}

// ============================================================================
// UdpSocket — Connectionless Datagram Socket
// ============================================================================
//
// Unlike TCP, a UDP socket has no connection state. Each send_to() specifies
// the destination independently. Incoming datagrams arrive in a queue with
// sender information attached (so the application knows who sent what).
//
// ============================================================================

/// A received UDP datagram with sender information.
#[derive(Debug, Clone)]
pub struct UdpDatagram {
    pub data: Vec<u8>,
    pub src_ip: Ipv4Address,
    pub src_port: u16,
}

pub struct UdpSocket {
    pub local_port: u16,
    recv_queue: Vec<UdpDatagram>,
}

impl UdpSocket {
    pub fn new(local_port: u16) -> Self {
        Self { local_port, recv_queue: Vec::new() }
    }

    /// Prepare a datagram for sending. Returns (header, data).
    pub fn send_to(&self, data: &[u8], dst_port: u16) -> (UdpHeader, Vec<u8>) {
        let header = UdpHeader::new(self.local_port, dst_port, 8 + data.len() as u16);
        (header, data.to_vec())
    }

    /// Deliver an incoming datagram to this socket's receive queue.
    pub fn deliver(&mut self, data: Vec<u8>, src_ip: Ipv4Address, src_port: u16) {
        self.recv_queue.push(UdpDatagram { data, src_ip, src_port });
    }

    /// Retrieve the next datagram from the queue, or None if empty.
    pub fn receive_from(&mut self) -> Option<UdpDatagram> {
        if self.recv_queue.is_empty() {
            None
        } else {
            Some(self.recv_queue.remove(0))
        }
    }

    /// Number of datagrams waiting in the queue.
    pub fn pending(&self) -> usize {
        self.recv_queue.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_header_serialize_deserialize() {
        let header = UdpHeader { src_port: 12345, dst_port: 53, length: 20, checksum: 0xABCD };
        let bytes = header.serialize();
        let restored = UdpHeader::deserialize(&bytes).unwrap();

        assert_eq!(restored.src_port, 12345);
        assert_eq!(restored.dst_port, 53);
        assert_eq!(restored.length, 20);
        assert_eq!(restored.checksum, 0xABCD);
    }

    #[test]
    fn test_header_8_bytes() {
        let header = UdpHeader::new(1000, 2000, 8);
        assert_eq!(header.serialize().len(), 8);
    }

    #[test]
    fn test_deserialize_too_short() {
        assert!(UdpHeader::deserialize(&[0u8; 7]).is_none());
        assert!(UdpHeader::deserialize(&[]).is_none());
    }

    #[test]
    fn test_default_checksum() {
        let header = UdpHeader::new(100, 200, 8);
        assert_eq!(header.checksum, 0);
    }

    #[test]
    fn test_send_to() {
        let sock = UdpSocket::new(5000);
        let (header, data) = sock.send_to(&[1, 2, 3, 4, 5], 53);

        assert_eq!(header.src_port, 5000);
        assert_eq!(header.dst_port, 53);
        assert_eq!(header.length, 13); // 8 + 5
        assert_eq!(data, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_deliver_and_receive() {
        let mut sock = UdpSocket::new(5000);
        sock.deliver(vec![10, 20, 30], [192, 168, 1, 1], 12345);

        let dgram = sock.receive_from().unwrap();
        assert_eq!(dgram.data, vec![10, 20, 30]);
        assert_eq!(dgram.src_ip, [192, 168, 1, 1]);
        assert_eq!(dgram.src_port, 12345);
    }

    #[test]
    fn test_receive_empty() {
        let mut sock = UdpSocket::new(5000);
        assert!(sock.receive_from().is_none());
    }

    #[test]
    fn test_fifo_ordering() {
        let mut sock = UdpSocket::new(5000);
        sock.deliver(vec![1], [10, 0, 0, 1], 1000);
        sock.deliver(vec![2], [10, 0, 0, 2], 2000);
        sock.deliver(vec![3], [10, 0, 0, 3], 3000);

        assert_eq!(sock.receive_from().unwrap().data, vec![1]);
        assert_eq!(sock.receive_from().unwrap().data, vec![2]);
        assert_eq!(sock.receive_from().unwrap().data, vec![3]);
        assert!(sock.receive_from().is_none());
    }

    #[test]
    fn test_pending() {
        let mut sock = UdpSocket::new(5000);
        assert_eq!(sock.pending(), 0);
        sock.deliver(vec![1], [0; 4], 100);
        sock.deliver(vec![2], [0; 4], 200);
        assert_eq!(sock.pending(), 2);
    }
}
